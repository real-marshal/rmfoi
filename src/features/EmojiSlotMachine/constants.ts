export const emojisMap = {
  '🐈': 'meow meow bitch',
  '🤦‍♀️': 'facepalm',
  '💀': 'dead',
  '🤡': 'oficially a clown',
  '🚀': 'to the moon!',
  '🤓': 'too nerdy',
  '🤯': 'mind fking blown',
  '🤑': '$$$$$$$$',
  '💩': 'shiiiit',
  '😈': 'so bad',
  '👊': 'nice one bro',
  '💪': 'gigachad',
  '💅': 'yas queen',
  '🦍': 'apes together strong',
  '💦': 'aaahhhhh',
  '🍑': 'nice booty',
  '🌶': 'damn so hot',
  '🗿': 'bruh',
  '🍆': 'wow so long',
  '📈': 'stonks go up',
  '💎': 'diamond hands',
  '🙈': 'nothing to see here',
  '🍻': 'cheers',
  '🏆': 'actually won the game',
  '🎉': 'congrats',
  '👑': 'king',
} as const

export type Emoji = keyof typeof emojisMap

export const emojis = Object.keys(emojisMap) as readonly Emoji[]

export const MIN_CHANCE = 1
export const MAX_CHANCE = 100

export enum EmojiSlotMachineMode {
  EASY = 'easy',
  STANDARD = 'standard',
  HARD = 'hard',
}

export const modeData = {
  [EmojiSlotMachineMode.EASY]: {
    chance: 40,
    chanceIncrement: 5,
  },
  [EmojiSlotMachineMode.STANDARD]: {
    chance: 20,
    chanceIncrement: 5,
  },
  [EmojiSlotMachineMode.HARD]: {
    chance: 10,
    chanceIncrement: 2.5,
  },
} as const

export const loseMessages = ['loser', 'meh', 'rekt', 'lost again?', 'fail'] as const
