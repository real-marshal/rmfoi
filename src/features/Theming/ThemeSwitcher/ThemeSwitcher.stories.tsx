import type { Meta, Story } from '@storybook/react'
import { ThemeSwitcher } from './ThemeSwitcher'

export default {
  title: 'Features/ThemeSwitcher',
  component: ThemeSwitcher,
} as Meta

const Template: Story = (args) => <ThemeSwitcher {...args} />

export const Basic = Template.bind({})
Basic.args = {}
