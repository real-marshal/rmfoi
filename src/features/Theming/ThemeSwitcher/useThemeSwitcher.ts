import { useEffect, useState } from 'react'
import { useMediaQuery } from '@/hooks'

const defaultThemeFallback = 'light'

export const useThemeSwitcher = (defaultTheme = 'system') => {
  const prefersDark = useMediaQuery('(prefers-color-scheme: dark)')
  const prefersLight = useMediaQuery('(prefers-color-scheme: light)')

  const persistedTheme = localStorage.getItem('theme')

  const [theme, setTheme] = useState(persistedTheme ?? defaultTheme)

  const actualTheme =
    theme === 'system'
      ? // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        (prefersDark && 'dark') || (prefersLight && 'light') || defaultThemeFallback
      : theme

  useEffect(() => {
    document.body.dataset['theme'] = actualTheme

    localStorage.setItem('theme', theme)
  }, [theme, prefersDark, prefersLight, actualTheme])

  return { theme, actualTheme, setTheme }
}
