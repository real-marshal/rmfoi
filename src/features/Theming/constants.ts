export enum ThemeValue {
  SYSTEM = 'system',
  DARK = 'dark',
  LIGHT = 'light',
}

export const SYSTEM_THEME_FALLBACK = ThemeValue.LIGHT

export type SystemThemeValue = Exclude<ThemeValue, ThemeValue.SYSTEM>

type ThemeKey =
  | 'background'
  | 'primary'
  | 'gradientBlue'
  | 'gradientBlueGreyed'
  | 'gradientGreenBlue'

export type Theme = Readonly<Record<ThemeKey, string>>
