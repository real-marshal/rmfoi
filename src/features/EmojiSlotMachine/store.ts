import store, { MakeCurrentDispatch, MakeCurrentState, MakeCurrentThunk } from '@/app/store'
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { Emoji, emojis, EmojiSlotMachineMode, modeData } from './constants'
import { getShuffledEmojiArrays, getWonEmojiIndexes, EmojiSwapData, tryToWin } from './utils'

const NAME = 'emojiSlotMachine' as const

interface EmojiSlotMachineState {
  mode: EmojiSlotMachineMode
  columns: number
  emojiArrays: Emoji[][]
  chance: number
  chanceIncrement: number
  minIndexToWin: number
  maxIndexToWin: number
  stats: {
    pulls: number
    wonEmojis: Partial<Record<Emoji, number>>
  }
}

// This slice will be injected asyncronously which is why we need separate types here
interface SliceState {
  [NAME]: EmojiSlotMachineState
}
export type CurrentRootState = MakeCurrentState<SliceState>
export type CurrentDispatch = MakeCurrentDispatch<SliceState>
// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CurrentThunk<ReturnType = void> = MakeCurrentThunk<SliceState, ReturnType>

export const useCurrentDispatch: () => CurrentDispatch = useDispatch
export const useCurrentSelector: TypedUseSelectorHook<CurrentRootState> = useSelector

const initialState: EmojiSlotMachineState = {
  mode: EmojiSlotMachineMode.STANDARD,
  columns: 3,
  emojiArrays: getShuffledEmojiArrays(3),
  minIndexToWin: 0,
  maxIndexToWin: emojis.length - 1,
  stats: {
    pulls: 0,
    wonEmojis: {},
  },
  ...modeData[EmojiSlotMachineMode.STANDARD],
}

/*

REDUCERS

*/

const { actions, reducer } = createSlice({
  name: NAME,
  initialState,
  reducers: {
    initialize: (
      state,
      action: PayloadAction<
        Partial<Pick<EmojiSlotMachineState, 'columns' | 'mode' | 'minIndexToWin' | 'maxIndexToWin'>>
      >
    ) => {
      Object.assign(state, action.payload)
    },
    setColumns: (state, action: PayloadAction<number>) => {
      state.columns = action.payload
    },
    setEmojiArrays: (state, action: PayloadAction<EmojiSlotMachineState['emojiArrays']>) => {
      state.emojiArrays = action.payload
    },
    win: (state, action: PayloadAction<Emoji>) => {
      state.stats.pulls++

      Object.assign(state, modeData[state.mode])

      const wonEmoji = state.stats.wonEmojis[action.payload]
      state.stats.wonEmojis[action.payload] = wonEmoji ? wonEmoji + 1 : 1
    },
    lose: (state) => {
      state.stats.pulls++
      state.chance += state.chanceIncrement
    },
    swapEmojis: (state, action: PayloadAction<EmojiSwapData[]>) => {
      const { emojiArrays } = state

      action.payload.forEach(
        ({ arrayIndex, emojiInd1, emojiInd2 }) =>
          ([emojiArrays[arrayIndex]![emojiInd1]!, emojiArrays[arrayIndex]![emojiInd2]!] = [
            emojiArrays[arrayIndex]![emojiInd2]!,
            emojiArrays[arrayIndex]![emojiInd1]!,
          ])
      )
    },
    updateMode: (state, action: PayloadAction<EmojiSlotMachineMode>) => {
      state.mode = action.payload
      Object.assign(state, modeData[state.mode])
    },
  },
})

export { actions }

store.injectReducer(NAME, reducer)

/*

THUNKS

*/

// shuffleEmojiArrays has impure RNG logic so it shouldn't be in the reducer
export const shuffleEmojiArrays = (): CurrentThunk => (dispatch, getState) => {
  const { columns } = getState()[NAME]
  const shuffledEmojis = getShuffledEmojiArrays(columns)

  dispatch(actions.setEmojiArrays(shuffledEmojis))
}

export const updateColumns =
  (columns: number): CurrentThunk =>
  (dispatch) => {
    dispatch(actions.setColumns(columns))
    dispatch(shuffleEmojiArrays())
  }

export const evaluateNewPull =
  (): CurrentThunk<{ wonEmojiIndexes: number[]; wonEmoji: Emoji } | undefined> =>
  (dispatch, getState) => {
    dispatch(shuffleEmojiArrays())

    const { chance, minIndexToWin, maxIndexToWin, emojiArrays } = getState()[NAME]
    const wonEmoji = tryToWin({ chance, minIndexToWin, maxIndexToWin })

    if (!wonEmoji) {
      dispatch(actions.lose())
      return
    }

    dispatch(actions.win(wonEmoji))

    const { wonEmojiIndexes, emojisToSwap } = getWonEmojiIndexes({
      emojiArrays,
      wonEmoji,
      minIndexToWin,
      maxIndexToWin,
    })
    dispatch(actions.swapEmojis(emojisToSwap))

    return { wonEmojiIndexes, wonEmoji }
  }

/*

SELECTORS

*/

export const selectEmojiArrays = (state: CurrentRootState) => state[NAME].emojiArrays

export const selectMode = (state: CurrentRootState) => state[NAME].mode

export const selectChance = (state: CurrentRootState) => state[NAME].chance

export const selectChanceIncrement = (state: CurrentRootState) => state[NAME].chanceIncrement

export const selectStats = (state: CurrentRootState) => state[NAME].stats

export const selectGeneralStats = createSelector(
  selectStats,
  selectChance,
  selectChanceIncrement,
  ({ pulls, wonEmojis }, chance, chanceIncrement) => {
    const wins = Object.values(wonEmojis).reduce((result, wins) => result + wins, 0)

    return {
      pulls,
      wins,
      loses: pulls - wins,
      WR: pulls ? wins / pulls : 0,
      chance,
      chanceIncrement,
    }
  }
)

export const selectdEmojiStats = createSelector(selectStats, ({ wonEmojis }) => {
  const wonEmojisNumber = Object.keys(wonEmojis).length

  return {
    wonEmojisNumber,
    leftEmojis: emojis.length - wonEmojisNumber,
    wonEmojis,
  }
})
