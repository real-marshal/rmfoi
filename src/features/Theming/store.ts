import type { RootState } from '@/app/store'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { SystemThemeValue, SYSTEM_THEME_FALLBACK, ThemeValue } from './constants'
import storage from 'redux-persist/lib/storage'
import { PersistConfig, persistReducer } from 'redux-persist'

const NAME = 'theme' as const
interface ThemeState {
  currentTheme: ThemeValue
  systemTheme: SystemThemeValue
}

const initialState: ThemeState = {
  currentTheme: ThemeValue.SYSTEM,
  systemTheme: SYSTEM_THEME_FALLBACK,
}

const persistConfig: PersistConfig<ThemeState> = {
  key: NAME,
  storage,
} as const

/*

REDUCERS

*/

const { actions, reducer } = createSlice({
  name: NAME,
  initialState,
  reducers: {
    setCurrentTheme: (state, action: PayloadAction<ThemeValue>) => {
      state.currentTheme = action.payload
    },
    setSystemTheme: (state, action: PayloadAction<SystemThemeValue>) => {
      state.systemTheme = action.payload
    },
  },
})

export { actions, reducer }
export const themeReducer = persistReducer(persistConfig, reducer)

/*

SELECTORS

*/

export const selectCurrentTheme = (state: RootState) => state.theme.currentTheme

export const selectActualTheme = (state: RootState) =>
  state.theme.currentTheme === ThemeValue.SYSTEM
    ? state.theme.systemTheme
    : state.theme.currentTheme
