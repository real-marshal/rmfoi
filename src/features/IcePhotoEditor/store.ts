import store, { MakeCurrentDispatch, MakeCurrentState, MakeCurrentThunk } from '@/app/store'
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { AdjustmentKey, AdjustmentsState, applyAdjustments } from './adjustments'
import { PersistConfig, persistReducer, PersistedState } from 'redux-persist'
import { createIDBStorage } from '@/common/rp-storage-idb'

const NAME = 'icePhotoEditor' as const

interface AdjustmentChange {
  name: AdjustmentKey
  prevValue: Exclude<AdjustmentsState[AdjustmentKey], undefined>
  currentValue: Exclude<AdjustmentsState[AdjustmentKey], undefined>
}

interface HistoryElement {
  date: number
  adjustment: AdjustmentChange
}

interface IcePhotoEditorState {
  /**
   * The original image that is never modified
   */
  image?: ImageData
  /**
   * The image that is the result of all adjustments applied
   */
  currentImage?: ImageData
  historyStack: HistoryElement[]
  currentHistoryPointer: number
  loadedDate?: number
  filename?: string
}

// This slice will be injected asyncronously which is why we need separate types here
interface SliceState {
  [NAME]: IcePhotoEditorState & Partial<PersistedState>
}
export type CurrentRootState = MakeCurrentState<SliceState>
export type CurrentDispatch = MakeCurrentDispatch<SliceState>
// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CurrentThunk<ReturnType = void> = MakeCurrentThunk<SliceState, ReturnType>

export const useCurrentDispatch: () => CurrentDispatch = useDispatch
export const useCurrentSelector: TypedUseSelectorHook<CurrentRootState> = useSelector

const initialState: IcePhotoEditorState = {
  historyStack: [],
  currentHistoryPointer: -1,
}

const persistConfig: PersistConfig<IcePhotoEditorState> = {
  key: NAME,
  // localStorage doesn't have enough space for this store
  storage: createIDBStorage(NAME),
  serialize: false,
  deserialize: false,
} as const

/*

REDUCERS

*/

const { actions, reducer } = createSlice({
  name: NAME,
  initialState,
  reducers: {
    loadImage: (state, action: PayloadAction<{ imageData: ImageData; filename: string }>) => {
      Object.assign(state, initialState)
      state.currentImage = state.image = action.payload.imageData
      state.filename = action.payload.filename

      state.loadedDate = Date.now()
    },
    addHistoryElement: (state, action: PayloadAction<HistoryElement>) => {
      state.historyStack.push(action.payload)
      state.currentHistoryPointer = state.historyStack.length - 1
    },
    updateLastHistoryElement: (state, action: PayloadAction<HistoryElement>) => {
      const { adjustment: lastAdjustment } = state.historyStack.at(-1) ?? {}

      if (!lastAdjustment) return

      Object.assign(lastAdjustment, action.payload.adjustment)
    },
    setCurrentHistoryElement: (state, action: PayloadAction<number>) => {
      const historyElementIndex = state.historyStack.findIndex(
        ({ date }) => date === action.payload
      )

      state.currentHistoryPointer = historyElementIndex

      if (action.payload === state.loadedDate) {
        state.currentHistoryPointer = -1
      }
    },
    applyAdjustments: (state) => {
      if (!state.image) return

      const historyStack = state.historyStack.slice(0, state.currentHistoryPointer + 1)

      const adjustmentsState = historyStack.reduce<AdjustmentsState>(
        (result, { adjustment }) => ({
          ...result,
          [adjustment.name]: adjustment.currentValue,
        }),
        {}
      )

      // eslint-disable-next-line unicorn/prefer-spread
      const newImageData = applyAdjustments(state.image.data.slice(0), adjustmentsState)

      state.currentImage = new ImageData(newImageData, state.image.width, state.image.height, {
        colorSpace: state.image.colorSpace,
      })
    },
  },
})

export { actions }

store.injectReducer(NAME, persistReducer(persistConfig, reducer))

/*

THUNKS

*/

export const updateAdjustments =
  (adjustment: AdjustmentChange): CurrentThunk =>
  (dispatch, getState) => {
    const { historyStack } = getState()[NAME]

    const { adjustment: lastAdjustment } = historyStack.at(-1) ?? {}
    const { adjustment: prevAdjustment } =
      historyStack.findLast(
        (historyElement) =>
          historyElement.adjustment.name === adjustment.name &&
          historyElement.adjustment !== lastAdjustment
      ) ?? {}

    if (prevAdjustment) {
      adjustment.prevValue = prevAdjustment.currentValue
    }

    if (lastAdjustment && lastAdjustment.name === adjustment.name) {
      dispatch(actions.updateLastHistoryElement({ date: Date.now(), adjustment }))
    } else {
      dispatch(actions.addHistoryElement({ date: Date.now(), adjustment }))
    }

    dispatch(actions.applyAdjustments())
  }

/*

SELECTORS

*/

export const selectImage = (state: CurrentRootState) => state[NAME].image

export const selectCurrentImage = (state: CurrentRootState) => state[NAME].currentImage

export const selectHistoryStack = (state: CurrentRootState) => state[NAME].historyStack

export const selectLoadedDate = (state: CurrentRootState) => state[NAME].loadedDate

export const selectFilename = (state: CurrentRootState) => state[NAME].filename

export const selectPersistRehydrated = (state: CurrentRootState) => state[NAME]._persist?.rehydrated

export const selectCurrentHistoryPointer = (state: CurrentRootState) =>
  state[NAME].currentHistoryPointer

export const selectHistory = createSelector(
  selectHistoryStack,
  selectLoadedDate,
  (historyStack, loadedDate) => [
    ...(loadedDate
      ? [
          {
            date: loadedDate,
            adjustment: {
              name: 'Image Loaded',
              prevValue: '',
              currentValue: '',
            },
          },
        ]
      : []),
    ...historyStack,
  ]
)

export const selectAdjustmentsStateStrings = createSelector(
  selectHistoryStack,
  selectCurrentHistoryPointer,
  (historyStack, currentHistoryPointer) =>
    historyStack.slice(0, currentHistoryPointer + 1).reduce<Partial<Record<AdjustmentKey, string>>>(
      (result, { adjustment }) => ({
        ...result,
        [adjustment.name]: adjustment.currentValue.toString(),
      }),
      {}
    )
)
