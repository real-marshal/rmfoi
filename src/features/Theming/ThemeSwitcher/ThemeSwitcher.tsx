import { capitalize } from '@/utils'
import { List } from '@/components'
import { useRef, useState } from 'react'
import { BsMoon, BsSun } from 'react-icons/bs'
import styled from '@emotion/styled'
import { dt } from '../design-tokens'
import { useOutsideClick } from '@/hooks'
import { ThemeValue } from '../constants'
import { useAppDispatch, useAppSelector } from '@/app/store'
import { actions, selectActualTheme, selectCurrentTheme } from '../store'

interface ThemeSwitcherProps {
  className?: string
}

const IconButton = styled.button`
  color: ${dt.theme.primary};
  border: 2px solid ${dt.theme.primary};
  border-radius: 10px;
  cursor: pointer;
  font-size: 2rem;
  background: ${dt.theme.background};
  padding: ${dt.padding.md};
  margin: 0;
  display: grid;
  place-items: center;

  &:hover {
    background: ${dt.theme.primary};
    color: ${dt.theme.background};
  }
`

const ThemeSwitcher = ({ className }: ThemeSwitcherProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const ref = useRef<HTMLDivElement>(null)

  useOutsideClick(ref, () => setIsOpen(false))

  const currentTheme = useAppSelector(selectCurrentTheme)
  const actualTheme = useAppSelector(selectActualTheme)

  const dispatch = useAppDispatch()

  return (
    <div className={className} ref={ref}>
      <IconButton onClick={() => setIsOpen(!isOpen)}>
        {actualTheme === ThemeValue.DARK ? <BsMoon /> : <BsSun />}
      </IconButton>
      {isOpen && (
        <List align='right'>
          {Object.values(ThemeValue).map((theme) => (
            <List.ListItem
              key={theme}
              selected={currentTheme === theme}
              onClick={() => {
                dispatch(actions.setCurrentTheme(theme))
                setIsOpen(false)
              }}
              value={theme}
            >
              {capitalize(theme)}
            </List.ListItem>
          ))}
        </List>
      )}
    </div>
  )
}

export { ThemeSwitcher }
export type { ThemeSwitcherProps }
