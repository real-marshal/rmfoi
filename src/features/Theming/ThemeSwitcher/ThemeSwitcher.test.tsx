import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ThemeValue } from '../constants'
import { actions } from '../store'
import { ThemeSwitcher } from './ThemeSwitcher'

const mockLight = ThemeValue.LIGHT
const mockDispatch = jest.fn()

jest.mock('@/app/store', () => ({
  useAppSelector: () => mockLight,
  useAppDispatch: () => mockDispatch,
}))

describe('ThemeSwitcher component', () => {
  it('renders the button to change theme', () => {
    const view = render(<ThemeSwitcher />)
    expect(view.baseElement).toMatchSnapshot()
  })

  it('opens the menu to choose a theme when the button is clicked and dispatches action on theme click', async () => {
    const user = userEvent.setup()
    const view = render(<ThemeSwitcher />)

    await user.click(screen.getByRole('button'))

    expect(view.baseElement).toMatchSnapshot()

    await user.click(screen.getByText(/dark/i))

    expect(mockDispatch).toHaveBeenCalledWith({
      type: actions.setCurrentTheme.type,
      payload: ThemeValue.DARK,
    })
  })
})
