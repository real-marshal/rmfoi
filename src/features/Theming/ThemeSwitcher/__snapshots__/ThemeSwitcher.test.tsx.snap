// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher component opens the menu to choose a theme when the button is clicked and dispatches action on theme click 1`] = `
.emotion-0 {
  color: var(--dt-theme-primary);
  border: 2px solid var(--dt-theme-primary);
  border-radius: 10px;
  cursor: pointer;
  font-size: 2rem;
  background: var(--dt-theme-background);
  padding: var(--dt-padding-md);
  margin: 0;
  display: grid;
  place-items: center;
}

.emotion-0:hover {
  background: var(--dt-theme-primary);
  color: var(--dt-theme-background);
}

.emotion-2 {
  background: linear-gradient(var(--dt-theme-background), var(--dt-theme-background)) padding-box,var(--dt-theme-gradientBlue);
  border: 4px solid transparent;
  border-radius: 10px;
  z-index: var(--dt-z-menu);
  padding: 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: absolute;
  right: 0;
}

.emotion-4 {
  color: var(--dt-theme-primary);
  background-color: var(--dt-theme-background);
  list-style: none;
  padding: var(--dt-padding-lg) var(--dt-padding-xl);
  cursor: pointer;
}

.emotion-4:first-of-type {
  border-radius: 7px 7px 0 0;
}

.emotion-4:last-of-type {
  border-radius: 0 0 7px 7px;
}

.emotion-4:hover {
  color: var(--dt-theme-background);
  background-color: var(--dt-theme-primary);
}

.emotion-8 {
  color: var(--dt-theme-primary);
  background-color: var(--dt-theme-background);
  list-style: none;
  padding: var(--dt-padding-lg) var(--dt-padding-xl);
  cursor: pointer;
  color: var(--dt-theme-background);
  background-color: var(--dt-theme-primary);
}

.emotion-8:first-of-type {
  border-radius: 7px 7px 0 0;
}

.emotion-8:last-of-type {
  border-radius: 0 0 7px 7px;
}

.emotion-8:hover {
  color: var(--dt-theme-background);
  background-color: var(--dt-theme-primary);
}

<body>
  <div>
    <div>
      <button
        class="emotion-0 emotion-1"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 11a3 3 0 1 1 0-6 3 3 0 0 1 0 6zm0 1a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"
          />
        </svg>
      </button>
      <ul
        class="emotion-2 emotion-3"
      >
        <li
          class="emotion-4 emotion-5"
          value="system"
        >
          System
        </li>
        <li
          class="emotion-4 emotion-5"
          value="dark"
        >
          Dark
        </li>
        <li
          class="emotion-8 emotion-5"
          value="light"
        >
          Light
        </li>
      </ul>
    </div>
  </div>
</body>
`;

exports[`ThemeSwitcher component renders the button to change theme 1`] = `
.emotion-0 {
  color: var(--dt-theme-primary);
  border: 2px solid var(--dt-theme-primary);
  border-radius: 10px;
  cursor: pointer;
  font-size: 2rem;
  background: var(--dt-theme-background);
  padding: var(--dt-padding-md);
  margin: 0;
  display: grid;
  place-items: center;
}

.emotion-0:hover {
  background: var(--dt-theme-primary);
  color: var(--dt-theme-background);
}

<body>
  <div>
    <div>
      <button
        class="emotion-0 emotion-1"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 11a3 3 0 1 1 0-6 3 3 0 0 1 0 6zm0 1a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"
          />
        </svg>
      </button>
    </div>
  </div>
</body>
`;
